vm-virtio/src/vhost_user/net.rs

```diff
@@ -68,14 +69,15 @@ impl<'a> Net {
 
         // Get features from backend, do negotiation to get a feature collection which
         // both VMM and backend support.
-        let backend_features = vhost_user_net
+        let mut backend_features = vhost_user_net
```

Comment:
This is not the right approach. We don't need to duplicate everything through `backend_features` variable.
You need to add `avail_features |= 1 << virtio_net::VIRTIO_NET_F_MAC;` earlier in the code, and mask it with the `backend_features`. You should never use `backend_features` as the input for other functions.
And you need to do the check for the protocol features with `avail_features` (after it has been masked with `backend_features`).

Reply:
Sorry, I do not quit get your point about "You need to add avail_features |= 1 << virtio_net::VIRTIO_NET_F_MAC; earlier in the code, and mask it with the backend_features", do you mean to send the avail_features with VIRTIO_NET_F_MAC to backend? If ovs+dpdk get features it does not support, it will stop handling with error like "invalid features received". This is the issue encountered while running test case with the current CH code base.

Reply:
No, I'm saying to add `virtio_net::VIRTIO_NET_F_MAC` between line 62 and 63. Then, read the backend features by doing a get_features. Then mask the available_features with the backend_features, this will remove from available_features what is not supported by the backend.

And later when the driver will read the available_features, based on them, it will properly set the acked_features.
The driver will not set some features that are not exposed through available_features, that's why you don't need to save backend_features to mask them with the acked_features.

Reply:
Ok, git it. But if in that case, the avail_features does not have virtio_net::VIRTIO_NET_F_MAC at all when guest OS call get_features to obtain device_features, and then device_features & driver_features, virtio_net in guest will have a feature set without VIRTIO_NET_F_MAC finally, it will just generate a MAC random, rather than read from config space.

Reply:
Oh interesting, now I understand the issue you are running into. So you're saying that we can have a set of features between device and driver different than the set between backend and device.
I'm okay with the patch then! I think there's one line that can be removed though.

Reply:
Exactly! Cool!